orttraining/orttraining/training_ops/cuda/nn/conv_grad.cc

```diff
@@ -4,114 +4,329 @@
-      args.c_desc,
-      args.w_desc,
-      algo,
-      sz);
+cudnnStatus_t GetWorkspaceSize(const ConvArgs& args, T_BwdDataAlgo algo, size_t* workspace_size) {
+  return cudnnGetConvolutionBackwardDataWorkspaceSize(args.handle, args.w_desc, args.y_tensor, args.conv_desc,
+                                                      args.x_tensor, algo, workspace_size);
 }
 
-// TODO: we can cache the descriptors, and only update if the input shape changes
-template <typename T>
-Status ConvGrad<T>::PrepareArgs(const Tensor& input, const Tensor& output, const Tensor& weight, const Tensor* bias) const {
-  const TensorShape& i_shape = input.Shape();
-  std::vector<int64_t> i_dims = i_shape.GetDims();
-
-  const TensorShape& o_shape = output.Shape();
-  std::vector<int64_t> o_dims = o_shape.GetDims();
+cudnnStatus_t GetWorkspaceSize(const ConvArgs& args, T_BwdFilterAlgo algo, size_t* workspace_size) {
+  return cudnnGetConvolutionBackwardFilterWorkspaceSize(args.handle, args.x_tensor, args.y_tensor, args.conv_desc,
+                                                        args.w_desc, algo, workspace_size);
+}
 
-  const TensorShape& w_shape = weight.Shape();
-  std::vector<int64_t> w_dims = w_shape.GetDims();
+template <typename T_Algo>
+size_t GetMaxWorkspaceSize(const ConvArgs& args, const T_Algo* algo, int n_algo) {
+  // Calling cudaMemGetInfo is not ideal, but our cuda allocator doesn't have a way to get this info.
+  size_t free, total;
+  CUDA_CALL_THROW(cudaMemGetInfo(&free, &total));
+  // Assuming 10% of fragmentation.
+  free = static_cast<size_t>(static_cast<double>(free) * 0.9);
+  size_t max_workspace_size = 0;
+  for (int i = 0; i < n_algo; i++) {
+    cudnnStatus_t status;
+    size_t workspace_size;
+    status = GetWorkspaceSize(args, algo[i], &workspace_size);
+    if (CUDNN_STATUS_SUCCESS != status || workspace_size == 0 || workspace_size < max_workspace_size ||
+        workspace_size > free)
+      continue;
+    max_workspace_size = workspace_size;
+  }
 
-  // Update Attributes
-  ORT_RETURN_IF_ERROR(conv_attrs_.ValidateInputShape(&input, &weight));
+  return max_workspace_size;
+}
 
-  std::vector<int64_t> kernel_shape;
-  ORT_RETURN_IF_ERROR(conv_attrs_.ComputeKernelShape(w_shape, kernel_shape));
-  auto rank = kernel_shape.size();
+template <typename T_Perf>
+std::vector<T_Perf> GetValidAlgorithms(T_Perf* perf_results, int n_algo) {
+  std::vector<T_Perf> result;
+  result.reserve(n_algo);
+  for (int i = 0; i < n_algo; i++) {
+    T_Perf perf = perf_results[i];
+    if (perf.status == CUDNN_STATUS_SUCCESS) {
+      result.emplace_back(perf);
+    }
+  }
+  ORT_ENFORCE(result.size() > 0, "No valid convolution algorithms available in CuDNN");
+  return result;
+}
 
-  std::vector<int64_t> pads(conv_attrs_.pads);
-  if (pads.empty()) {
-    pads.resize(rank * 2, 0);
+struct ConvParamsHash {
+  size_t operator()(const ConvParams& conv_params) const {
+    auto ptr = reinterpret_cast<const uint8_t*>(&conv_params);
```

Comment:
static_assert(std::is_pod<ConvParams>::value, "ConvParams is not POD"); 

to prevent accidental future change of ConvParams